src/Microsoft.AspNetCore.Mvc/MvcServiceCollectionExtensions.cs

```diff
@@ -29,6 +31,8 @@ public static IMvcBuilder AddMvc(this IServiceCollection services)
                 throw new ArgumentNullException(nameof(services));
             }
 
+            services.AddSingleton<IApplicationNameProvider>(new HostingApplicationNameProvider(services));
```

Comment:
I'm not super pumped that we add this in AddMvc instead of AddMvcCore but doing it in core would mean Core takes a dependency on Localization, which I feel is even less desirable.